### ISSUE TITLE
Database has encountered some problem, please perform necessary action (tx recovery/restart)

### ISSUE THREAD

**Issue by nateless** [neo4j/neo4j#8687]
- Neo4j version: 3.1.0
- Operating system: Ubuntu 16.10
- API/Driver: Cypher

We are doing realtime data import to neo, which can peak up to 100 ops. Last weekend it crashed, is it something we are doing wrong or it's neo's bug?

```
2017-01-21 03:56:53.216+0000 INFO  [o.n.k.NeoStoreDataSource] Rotating log version:20
2017-01-21 03:56:53.216+0000 INFO  [o.n.k.NeoStoreDataSource] Opened logical log [/var/lib/neo4j/data/databases/graph.db/neostore.transaction.db.21] version=21, lastTxId=2064659 (clean)
2017-01-21 03:56:53.217+0000 INFO  [o.n.k.NeoStoreDataSource] Finished rotating log version:20
2017-01-21 03:59:05.739+0000 INFO  [o.n.k.i.t.l.c.CheckPointerImpl] Check Pointing triggered by scheduler for time threshold [2072011]:  Starting check pointing...
2017-01-21 03:59:05.739+0000 INFO  [o.n.k.i.t.l.c.CheckPointerImpl] Check Pointing triggered by scheduler for time threshold [2072011]:  Starting store flush...
2017-01-21 03:59:05.788+0000 INFO  [o.n.k.i.s.c.CountsTracker] About to rotate counts store at transaction 2072015 to [/var/lib/neo4j/data/databases/graph.db/neostore.counts.db.b], from [/var/lib/neo4j/data/databases/graph.db/neostore.counts.db.a].
2017-01-21 03:59:05.788+0000 INFO  [o.n.k.i.s.c.CountsTracker] Successfully rotated counts store at transaction 2072015 to [/var/lib/neo4j/data/databases/graph.db/neostore.counts.db.b], from [/var/lib/neo4j/data/databases/graph.db/neostore.counts.db.a].
2017-01-21 03:59:39.316+0000 INFO  [o.n.k.i.t.l.c.CheckPointerImpl] Check Pointing triggered by scheduler for time threshold [2072011]:  Store flush completed
2017-01-21 03:59:39.316+0000 INFO  [o.n.k.i.t.l.c.CheckPointerImpl] Check Pointing triggered by scheduler for time threshold [2072011]:  Starting appending check point entry into the tx log...
2017-01-21 03:59:39.317+0000 INFO  [o.n.k.i.t.l.c.CheckPointerImpl] Check Pointing triggered by scheduler for time threshold [2072011]:  Appending check point entry into the tx log completed
2017-01-21 03:59:39.317+0000 INFO  [o.n.k.i.t.l.c.CheckPointerImpl] Check Pointing triggered by scheduler for time threshold [2072011]:  Check pointing completed
2017-01-21 03:59:39.317+0000 INFO  [o.n.k.i.t.l.p.LogPruningImpl] Log Rotation [21]:  Starting log pruning.
2017-01-21 03:59:39.317+0000 INFO  [o.n.k.i.t.l.p.LogPruningImpl] Log Rotation [21]:  Log pruning complete.
2017-01-21 04:04:39.321+0000 INFO  [o.n.k.i.t.l.c.CheckPointerImpl] Check Pointing triggered by scheduler for time threshold [2088744]:  Starting check pointing...
2017-01-21 04:04:39.321+0000 INFO  [o.n.k.i.t.l.c.CheckPointerImpl] Check Pointing triggered by scheduler for time threshold [2088744]:  Starting store flush...
2017-01-21 04:04:39.402+0000 INFO  [o.n.k.i.s.c.CountsTracker] About to rotate counts store at transaction 2088747 to [/var/lib/neo4j/data/databases/graph.db/neostore.counts.db.a], from [/var/lib/neo4j/data/databases/graph.db/neostore.counts.db.b].
2017-01-21 04:04:39.403+0000 INFO  [o.n.k.i.s.c.CountsTracker] Successfully rotated counts store at transaction 2088747 to [/var/lib/neo4j/data/databases/graph.db/neostore.counts.db.a], from [/var/lib/neo4j/data/databases/graph.db/neostore.counts.db.b].
2017-01-21 04:05:20.914+0000 INFO  [o.n.k.i.t.l.c.CheckPointerImpl] Check Pointing triggered by scheduler for time threshold [2088744]:  Store flush completed
2017-01-21 04:05:20.915+0000 INFO  [o.n.k.i.t.l.c.CheckPointerImpl] Check Pointing triggered by scheduler for time threshold [2088744]:  Starting appending check point entry into the tx log...
2017-01-21 04:05:20.915+0000 INFO  [o.n.k.i.t.l.c.CheckPointerImpl] Check Pointing triggered by scheduler for time threshold [2088744]:  Appending check point entry into the tx log completed
2017-01-21 04:05:20.915+0000 INFO  [o.n.k.i.t.l.c.CheckPointerImpl] Check Pointing triggered by scheduler for time threshold [2088744]:  Check pointing completed
2017-01-21 04:05:20.915+0000 INFO  [o.n.k.i.t.l.p.LogPruningImpl] Log Rotation [21]:  Starting log pruning.
2017-01-21 04:05:20.916+0000 INFO  [o.n.k.i.t.l.p.LogPruningImpl] Log Rotation [21]:  Log pruning complete.
2017-01-21 04:09:29.915+0000 ERROR [o.n.k.i.DatabaseHealth] Database panic: Database has encountered some problem, please perform necessary action (tx recovery/restart) Failed to flush index updates prior to applying schema change
java.io.IOException: Failed to flush index updates prior to applying schema change
	at org.neo4j.kernel.impl.transaction.command.IndexBatchTransactionApplier.applyIndexUpdates(IndexBatchTransactionApplier.java:98)
	at org.neo4j.kernel.impl.transaction.command.IndexBatchTransactionApplier.close(IndexBatchTransactionApplier.java:116)
	at org.neo4j.kernel.impl.api.BatchTransactionApplierFacade.close(BatchTransactionApplierFacade.java:70)
	at org.neo4j.kernel.impl.storageengine.impl.recordstorage.RecordStorageEngine.apply(RecordStorageEngine.java:354)
	at org.neo4j.kernel.impl.api.TransactionRepresentationCommitProcess.applyToStore(TransactionRepresentationCommitProcess.java:78)
	at org.neo4j.kernel.impl.api.TransactionRepresentationCommitProcess.commit(TransactionRepresentationCommitProcess.java:51)
	at org.neo4j.kernel.impl.api.KernelTransactionImplementation.commit(KernelTransactionImplementation.java:608)
	at org.neo4j.kernel.impl.api.KernelTransactionImplementation.closeTransaction(KernelTransactionImplementation.java:484)
	at org.neo4j.kernel.api.KernelTransaction.close(KernelTransaction.java:135)
	at org.neo4j.server.rest.transactional.TransitionalTxManagementKernelTransaction.commit(TransitionalTxManagementKernelTransaction.java:97)
	at org.neo4j.server.rest.transactional.TransactionHandle.closeContextAndCollectErrors(TransactionHandle.java:250)
	at org.neo4j.server.rest.transactional.TransactionHandle.commit(TransactionHandle.java:158)
	at org.neo4j.server.rest.web.TransactionalService.lambda$executeStatementsAndCommit$1(TransactionalService.java:216)
	at com.sun.jersey.core.impl.provider.entity.StreamingOutputProvider.writeTo(StreamingOutputProvider.java:71)
	at com.sun.jersey.core.impl.provider.entity.StreamingOutputProvider.writeTo(StreamingOutputProvider.java:57)
	at com.sun.jersey.spi.container.ContainerResponse.write(ContainerResponse.java:302)
	at com.sun.jersey.server.impl.application.WebApplicationImpl._handleRequest(WebApplicationImpl.java:1510)
	at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1419)
	at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1409)
	at com.sun.jersey.spi.container.servlet.WebComponent.service(WebComponent.java:409)
	at com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:558)
	at com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:733)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:808)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1669)
	at org.neo4j.server.rest.dbms.AuthorizationEnabledFilter.doFilter(AuthorizationEnabledFilter.java:122)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at org.neo4j.server.rest.web.CollectUserAgentFilter.doFilter(CollectUserAgentFilter.java:69)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:585)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:221)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:52)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:497)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:310)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:540)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.util.concurrent.ExecutionException: java.lang.OutOfMemoryError: unable to create new native thread
	at org.neo4j.concurrent.WorkSync.checkFailure(WorkSync.java:135)
	at org.neo4j.concurrent.WorkSync.apply(WorkSync.java:91)
	at org.neo4j.kernel.impl.transaction.command.IndexBatchTransactionApplier.applyIndexUpdates(IndexBatchTransactionApplier.java:94)
	... 44 more
Caused by: java.lang.OutOfMemoryError: unable to create new native thread
	at java.lang.Thread.start0(Native Method)
	at java.lang.Thread.start(Thread.java:714)
	at org.apache.lucene.index.ConcurrentMergeScheduler.merge(ConcurrentMergeScheduler.java:517)
	at org.apache.lucene.index.IndexWriter.maybeMerge(IndexWriter.java:1931)
	at org.apache.lucene.index.IndexWriter.getReader(IndexWriter.java:455)
	at org.apache.lucene.index.StandardDirectoryReader.doOpenFromWriter(StandardDirectoryReader.java:286)
	at org.apache.lucene.index.StandardDirectoryReader.doOpenIfChanged(StandardDirectoryReader.java:261)
	at org.apache.lucene.index.StandardDirectoryReader.doOpenIfChanged(StandardDirectoryReader.java:251)
	at org.apache.lucene.index.DirectoryReader.openIfChanged(DirectoryReader.java:137)
	at org.apache.lucene.search.SearcherManager.refreshIfNeeded(SearcherManager.java:154)
	at org.apache.lucene.search.SearcherManager.refreshIfNeeded(SearcherManager.java:58)
	at org.apache.lucene.search.ReferenceManager.doMaybeRefresh(ReferenceManager.java:176)
	at org.apache.lucene.search.ReferenceManager.maybeRefreshBlocking(ReferenceManager.java:253)
	at org.neo4j.kernel.api.impl.index.partition.WritableIndexPartition.maybeRefreshBlocking(WritableIndexPartition.java:76)
	at org.neo4j.kernel.api.impl.index.AbstractLuceneIndex.maybeRefreshPartition(AbstractLuceneIndex.java:296)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:291)
	at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:731)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinTask.doInvoke(ForkJoinTask.java:401)
	at java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:734)
	at java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:160)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:174)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:583)
	at org.neo4j.kernel.api.impl.index.AbstractLuceneIndex.maybeRefreshBlocking(AbstractLuceneIndex.java:284)
	at org.neo4j.kernel.api.impl.index.WritableAbstractDatabaseIndex.maybeRefreshBlocking(WritableAbstractDatabaseIndex.java:197)
	at org.neo4j.kernel.api.impl.schema.LuceneIndexAccessor$LuceneIndexUpdater.close(LuceneIndexAccessor.java:158)
	at org.neo4j.kernel.impl.api.index.updater.UpdateCountingIndexUpdater.close(UpdateCountingIndexUpdater.java:55)
	at org.neo4j.kernel.impl.api.index.ContractCheckingIndexProxy$1.close(ContractCheckingIndexProxy.java:104)
	at org.neo4j.kernel.impl.api.index.IndexUpdaterMap.close(IndexUpdaterMap.java:84)
	at org.neo4j.kernel.impl.api.index.IndexingService.apply(IndexingService.java:485)
	at org.neo4j.kernel.impl.api.index.IndexingService.apply(IndexingService.java:418)
	at org.neo4j.kernel.impl.transaction.command.IndexUpdatesWork.apply(IndexUpdatesWork.java:63)
	at org.neo4j.kernel.impl.transaction.command.IndexUpdatesWork.apply(IndexUpdatesWork.java:42)
	at org.neo4j.concurrent.WorkSync.doSynchronizedWork(WorkSync.java:184)
	at org.neo4j.concurrent.WorkSync.tryDoWork(WorkSync.java:110)
	... 46 more
```

**Comment by jexp** [comment]
Can you share more of your system configuration.

Total Memory, heap settings, what driver do you use and how (do you close results/statements).
More details on the queries you run.
Perhaps your full logfile would be helpful as it contains forensics.

Thanks

**Comment by nateless** [comment]
Total Ram 256GB
dbms.memory.pagecache.size=160g
dbms.memory.heap.initial_size=90000
dbms.memory.heap.max_size=90000

We use REST API to send cypher commands, it's a json data, which is being parsed with cypher. There's no logs in application server regarding this error, only that previous request successfully executed and next one that server is unreachable.

This one from neo4j.log

```
2016-12-23 12:13:09.785+0000 ERROR Client triggered an unexpected error [UnknownError]: This transaction has already been completed.. See debug.log for more details, reference 98a70d42-4f0b-4d7d-8dba-335736ed1733.
2017-01-20 08:59:17.194+0000 ERROR Client triggered an unexpected error [UnknownError]: The statement has been closed.. See debug.log for more details, reference 9113d249-6eaa-4209-b937-f81361412ae1.
2017-01-20 08:59:17.211+0000 ERROR Client triggered an unexpected error [UnknownError]: This transaction has already been completed.. See debug.log for more details, reference b142327c-9007-429a-a338-6fa793784d01.
2017-01-20 08:59:17.215+0000 ERROR Client triggered an unexpected error [UnknownError]: null. See debug.log for more details, reference 0f35bda2-aa68-4cd8-a7e3-ef65acae9a39.
2017-01-20 09:01:09.632+0000 ERROR Client triggered an unexpected error [UnknownError]: null. See debug.log for more details, reference 11eb08e0-0fd7-4d61-856b-9ac390ef2973.
2017-01-20 09:01:09.654+0000 ERROR Client triggered an unexpected error [UnknownError]: null. See debug.log for more details, reference 26551dbd-e5d6-435f-ac9a-cfc7820281c0.
2017-01-20 09:01:09.660+0000 ERROR Client triggered an unexpected error [UnknownError]: null. See debug.log for more details, reference 88cd968b-c1f2-47da-8d93-5a3a2244be47.
2017-01-20 09:01:09.684+0000 ERROR Client triggered an unexpected error [UnknownError]: null. See debug.log for more details, reference fa0aac56-bcbb-4307-bd93-c1505e277577.
2017-01-20 09:01:09.684+0000 ERROR Client triggered an unexpected error [UnknownError]: null. See debug.log for more details, reference a6baecda-cca5-4df0-bd46-dd97faac1a13.
2017-01-20 09:01:09.684+0000 ERROR Client triggered an unexpected error [UnknownError]: null. See debug.log for more details, reference 53f4ee60-3399-406a-982f-35023539cca8.
2017-01-20 09:01:09.686+0000 ERROR Client triggered an unexpected error [UnknownError]: null. See debug.log for more details, reference b6aebeaf-0183-4b1c-9859-eb2fd18eaf19.
2017-01-20 09:01:09.692+0000 ERROR Client triggered an unexpected error [UnknownError]: null. See debug.log for more details, reference 132f529a-a36b-4252-ab75-5b5be97a6990.
2017-01-20 09:01:09.852+0000 ERROR Client triggered an unexpected error [UnknownError]: The statement has been closed.. See debug.log for more details, reference 4757f99d-47b1-418c-958f-a025a3961f13.
2017-01-20 09:01:10.320+0000 ERROR Client triggered an unexpected error [UnknownError]: The statement has been closed.. See debug.log for more details, reference 387a3c6d-838e-4920-a1a7-c58257d76af5.
2017-01-20 09:06:23.136+0000 ERROR Client triggered an unexpected error [UnknownError]: null. See debug.log for more details, reference f58148a6-7f09-472d-bd0d-de339df193b4.
2017-01-20 09:06:23.143+0000 ERROR Client triggered an unexpected error [UnknownError]: null. See debug.log for more details, reference 0711d5e4-74d1-4e54-adb2-2d6f77ba4275.
2017-01-20 09:06:23.173+0000 ERROR Client triggered an unexpected error [UnknownError]: The statement has been closed.. See debug.log for more details, reference 70112e82-b980-43f1-a6eb-3d88409bd191.
2017-01-20 09:08:48.009+0000 ERROR Client triggered an unexpected error [UnknownError]: null. See debug.log for more details, reference 4f1f0a35-2983-4b7e-ab45-c01ef48a276a.
2017-01-20 09:13:30.396+0000 ERROR Client triggered an unexpected error [UnknownError]: null. See debug.log for more details, reference d3515438-a23c-4787-82b7-983370151b9e.
2017-01-20 09:18:34.163+0000 ERROR Client triggered an unexpected error [UnknownError]: null. See debug.log for more details, reference b9b909f7-c8e8-40c5-9262-7d3ad3d0a81f.
2017-01-20 09:27:27.419+0000 ERROR Client triggered an unexpected error [UnknownError]: null. See debug.log for more details, reference 92b8e61d-85e8-4607-b875-40c2a273be61.
2017-01-20 09:27:27.431+0000 ERROR Client triggered an unexpected error [UnknownError]: null. See debug.log for more details, reference 3d893f4f-2d04-459b-ad85-5f8e1a93d439.
2017-01-20 09:27:27.550+0000 ERROR Client triggered an unexpected error [UnknownError]: null. See debug.log for more details, reference 22fb5e87-87c9-4ad7-9a0b-9c7fac7ae804.
2017-01-20 09:27:27.553+0000 ERROR Client triggered an unexpected error [UnknownError]: The statement has been closed.. See debug.log for more details, reference 751a6ba3-71df-4c15-bd12-854073f10064.
2017-01-20 09:27:27.563+0000 ERROR Client triggered an unexpected error [UnknownError]: null. See debug.log for more details, reference 4066e034-62dc-490b-a5b0-d00ea84516e3.
2017-01-20 09:27:27.566+0000 ERROR Client triggered an unexpected error [UnknownError]: null. See debug.log for more details, reference 1364bebb-5842-4580-ba66-e4c26ac05655.
2017-01-20 09:27:27.599+0000 ERROR Client triggered an unexpected error [UnknownError]: null. See debug.log for more details, reference 160bb615-f8c3-4c2e-abc9-96f44377d8a1.
2017-01-20 09:27:38.806+0000 ERROR Client triggered an unexpected error [UnknownError]: The statement has been closed.. See debug.log for more details, reference d5911737-c992-44f8-86b9-c8e8e66a014f.
2017-01-20 09:27:39.287+0000 ERROR Client triggered an unexpected error [UnknownError]: null. See debug.log for more details, reference 580a912e-7089-416f-9a05-aac21baa9ee5.
2017-01-20 09:27:40.922+0000 ERROR Client triggered an unexpected error [UnknownError]: null. See debug.log for more details, reference b966bff2-97df-4dc4-bd28-a64c408b548f.
2017-01-21 04:09:29.920+0000 ERROR Client triggered an unexpected error [TransactionStartFailed]: Database has encountered some problem, please perform necessary action (tx recovery/restart). See debug.log for more details, reference 4f2344cf-95bf-4d90-9c33-276d0384ba7b.
```

**Comment by jexp** [comment]
How many CPUs (also vCpus) do you have, do you have any settings for max-server threads?

What's the data volume that you send / receive? And how many labels / indexes do you have?

Caused by: java.util.concurrent.ExecutionException: java.lang.OutOfMemoryError: unable to create new native thread
	at org.neo4j.concurrent.WorkSync.checkFailure(WorkSync.java:135)
	at org.neo4j.concurrent.WorkSync.apply(WorkSync.java:91)
	at org.neo4j.kernel.impl.transaction.command.IndexBatchTransactionApplier.applyIndexUpdates(IndexBatchTransactionApplier.java:94)
	... 44 more
Caused by: java.lang.OutOfMemoryError: unable to create new native thread
	at java.lang.Thread.start0(Native Method)
	at java.lang.Thread.start(Thread.java:714)
	at org.apache.lucene.index.ConcurrentMergeScheduler.merge(ConcurrentMergeScheduler.java:517)

**Comment by jexp** [comment]
and a thread dump when this happens would be great

**Comment by nateless** [comment]
We have 12 CPUs, it's dedicated machine. We do not have any settings for threads.

It's small, first we select whether we have this node by some attribute id, then insert new node and additional data with it, shouldn't be more than 10-20 new nodes and around 100 new relations per one such insert, something around 5-10kb in json data.

We have three nodes and three relations, 5 indexes, total db size 176.18 GiB 

How we can create the dump?

**Comment by jexp** [comment]
kill -3 <pid>
or jstack <pid>

**Comment by nateless** [comment]
Thanks, I will update this issue as soon as it will happen again.

**Comment by spacecowboy** [comment]
Your memory settings pretty much sum up to the total memory of the machine. I have found that to be a pretty strong guarantee for `OutOfMemoryException`. Especially on machines with lots of memory. You might want to try and change your memory settings so that 

dbms.memory.heap.max_size * 2 + dbms.memory.pagecache.size == machine RAM - OS-Ram-needs

**Comment by Aurelsicoko** [comment]
Hey guys, I encountered the exact same error when I tried to delete a lot of relationships/nodes (> 1M). Every time Neo4J crashed, and I need to restart it manually. I also tried to make the same scenario by deleting only 100K nodes/relationships and repeat this 10 times. But at the end, this is the same thing, the Neo4J is unusable.

What are the solutions?

**Comment by spacecowboy** [comment]
@Aurelsicoko as I mention above, do make sure your memory settings are good. e.g., don't consume the entire machine's RAM.

For deletes, you should also batch deletes in chunks of 10.000 or so.

**Comment by Aurelsicoko** [comment]
@spacecowboy Thanks for the recommendations, I will take a look at my configurations!

**Comment by tinwelint** [comment]
As for the initial issue, the lucene OOM... this is mostly fixed due to the native indexing that started in 3.3 and replaced the majority of indexing needs in 3.5. The native index uses a tiny fraction of heap memory compared to Lucene, virtually constant regardless of amount of writes.

@Aurelsicoko your issue is very different, even if it results in OOM too. Your issue is related to how transaction updates accumulates in heap memory. Something that is partly addressed in 3.5 by the possibility to take some of that memory off-heap. At some point the root cause of accumulating this much memory for changes will also be addressed, but I do believe that the problem in this issue has been addressed.
